Blocked domains flood gserver entries

Quix0r opened the pull request:
It has shown in the past that trolls from the infamous `gab.best` and also more recently `activitypub-troll.cf` domains have used a flaw in Friendica to flood the `gserver` table with millions (see [this link](https://friendica.myportal.social/display/e65e1095-1763-c947-a194-ab1699044817)) of useless records (tons of invalid and not findable sub domains) of records that just pile up and seem to never end.

With this change, I was finally able to block them from entering my and other's database tables. This PR also contains a change to to avoid redundant invocations of `Strings::normaliseLink()` which can safe some CPU cycles and speedup the code.

Quix0r committed:
Continued:
- some trolls managed to flood gserver with useless URLs. They can be blocked
  by domain blocking them, but still it floods gserver table with dead entries
- this hack tries to change that so they won't enter gserver at all. Let's hope
  these trolls as `activitypub-trolls.cf` learn a lesson (and get adults soon)

Signed-off-by: Roland Häder <roland@mxchange.org>

Quix0r committed:
Continued:
- just for consistency ...

MrPetovan commented on src/Model/GServer.php:
```diff
@@ -458,19 +458,30 @@ public static function setBlockedByUrl(string $url)
 	public static function setFailureByUrl(string $url)
 	{
-		$gserver = DBA::selectFirst('gserver', [], ['nurl' => Strings::normaliseLink($url)]);
+		$nurl = Strings::normaliseLink($url);
+
+		if (Network::isUrlBlocked($url)) {
+			Logger::info('Server domain is blocked', ['url' => $url]);
+			return;
+		} elseif (Network::isUrlBlocked($nurl)) {
+			Logger::info('Server domain is blocked', ['nurl' => $nurl]);
+			return;
+		}
```
Please move this block after the update condition, but before the insert call.

MrPetovan commented on src/Model/GServer.php:
```diff
@@ -560,6 +571,9 @@ private static function detect(string $url, string $network = '', bool $only_nod
 				self::detect($url, $network, $only_nodeinfo);
 			}
 			return false;
+		} elseif (Network::isUrlBlocked($url)) {
+			Logger::info('Server domain is blocked', ['url' => $url]);
+			return false;
```
`elseif` here is superfluous since there's a `return` in the parent condition.
```suggestion
		}

		if (Network::isUrlBlocked($url)) {
			Logger::info('Server domain is blocked', ['url' => $url]);
			return false;
```

Quix0r replied:
What do you mean by "parent condition"?

MrPetovan replied:
The condition where the `elseif` originated.

Quix0r replied:
I usually only make a 2nd `if()` block if there is code between the upper and the then added lower one:
````
if (someCondition) {
// Do something, like logging, closing/freeing resources
return false;
}

// Do other things:
doOtherCode();

if (otherCondition) {
// Do something, like logging, closing/freeing resources
return false;
}
````
Otherwise, when there is no code in between these two `if()` blocks, I merge them with `elseif()` or when they do the exact same thing, I merge them with `||`.

MrPetovan replied:
Let me provide a more actionable suggestion.
```suggestion
		if (!Strings::compareLink($url, $original_url)) {
			self::setFailureByUrl($original_url);
			if (!self::getID($url, true)) {
				self::detect($url, $network, $only_nodeinfo);
			}
			return false;
		}

		if (Network::isUrlBlocked($url)) {
			Logger::info('Server domain is blocked', ['url' => $url]);
			return false;
		}
```

These condition have nothing to do with each other because of the `return false`. If there was anything but a return a `elseif` would be warranted, but not in this case where it adds complexity for no benefit.

MrPetovan commented on src/Worker/UpdateServerPeers.php:
```diff
@@ -57,13 +57,13 @@ public static function execute(string $url)
 
 			++$total;
-			if (DBA::exists('gserver', ['nurl' => Strings::normaliseLink('http://' . $peer)])) {
+			if (DBA::exists('gserver', ['nurl' => Strings::normaliseLink('https://' . $peer)])) {
```
This is ineffective since `Strings::normaliseLink` whole purpose is to reduce `https` URL to `http`.
```suggestion
			if (DBA::exists('gserver', ['nurl' => 'http://' . $peer])) {
```

Quix0r replied:
So `nurl` contains the URL without any SSL, just like an namespace identifier you may know from XML files?

MrPetovan replied:
I'm not sure what you are referring to, but yes, normalized URLs are without SSL. We assume that the same content is served whether the same URL is accessed with or without SSL, so we normalize without SSL to prevent duplicate entries for contacts and servers.

Quix0r replied:
Okay, then I revert this part and take your suggestion.

MrPetovan commented on src/Worker/UpdateServerPeers.php:
```diff
@@ -57,13 +57,13 @@ public static function execute(string $url)
 		$added = 0;
 		foreach ($peers as $peer) {
-			if (Network::isUrlBlocked('http://' . $peer)) {
+			if (Network::isUrlBlocked('https://' . $peer)) {
```
We only match on the domain so far, so this change would be inconsequential, although it looks better?

Quix0r replied:
Okay, then. http:// isn't part of the validation.

MrPetovan replied:
Yes, but we still need a scheme to be able to take advantage of the URL parsing method used in `Network::isUrlBlocked`.

Quix0r commented:
````
DELETE FROM `gserver` WHERE `url` LIKE '%activitypub-troll.cf%' OR `url` LIKE '%gab.best%'

Query executed OK, 0 rows affected. (0.149 s)
````
So far really effective! Without this change, both the later one would have flooded again with even millions(!) of sub domains.

Quix0r committed:
Continued:
- moved if() block to suggested position by MrPetovan, for me I want to have all
  conditions checked at the start of the method, e.g. no unwanted null
  references or (in this case) if the URL is blacklisted
- normalized URLs are without SSL, means http://host/path/file.ext so they exist
  only once for contacts and servers (aka. instances)
- documented returned type `void`

MrPetovan commented on src/Model/GServer.php:
```diff
@@ -481,6 +474,14 @@ public static function setFailureByUrl(string $url)
 			return;
 		}
 
+		if (Network::isUrlBlocked($url)) {
+			Logger::info('Server domain is blocked', ['url' => $url]);
+			return;
+		} elseif (Network::isUrlBlocked($nurl)) {
+			Logger::info('Server domain is blocked', ['nurl' => $nurl]);
+			return;
+		}
```
The reason why I asked you to move these conditions later is that this is a method to update a potentially existing server record. If the record was created earlier (and not purged adding a new domain pattern to the block list) then we do want to keep it accurate, even if it is currently blocked.

On the other hand, we absolutely do not want to create a new record if the domain is blocked. So we need to allow the update case even if the domain is blocked.

annando replied:
We don't need this check since we check for blocked domains in the parent functions. After I deleted the server entries from my system several days ago, the current code worked perfectly and hadn't created any new of these entries.

MrPetovan replied:
Should we revert then?

Quix0r replied:
Here it didn't work perfectly. My `gserver` table was still flooded with +1.7 million records non-stop. Only when I added the code, it was finally solved. Okay, I can raise the log level to see if these messages there are showing up.

Quix0r replied:
Here is another report of being flooded by these trolls (without this PR): https://fika.grin.hu/display/9b3ec022-4463-d42e-0a21-6b3257115451

annando replied:
Yes, he is on 2022.10. We improved the server block checks with 2023.01.

annando commented on src/Model/GServer.php:
```diff
@@ -455,22 +455,34 @@ public static function setBlockedByUrl(string $url)
 		}
-		self::insert(['url' => $url, 'nurl' => Strings::normaliseLink($url),
+
+		if (Network::isUrlBlocked($url)) {
```
This block here is not needed. AFAIK we don't call this function at all for blocked servers. If we should, then we had to check earlier for blocked servers and should call the appropriate function from there.

annando commented on src/Model/GServer.php:
```diff
@@ -560,6 +572,9 @@ private static function detect(string $url, string $network = '', bool $only_nod
 				self::detect($url, $network, $only_nodeinfo);
 			}
 			return false;
+		} elseif (Network::isUrlBlocked($url)) {
```
The `detect` function is only called via the `check` function that checks for a block in the very first lines. So it shouldn't happen at all that a blocked server could slip through.

Quix0r replied:
Thank you for taking care of this.

annando commented on src/Worker/UpdateServerPeers.php:
```diff
@@ -57,13 +57,13 @@ public static function execute(string $url)
 
 			++$total;
-			if (DBA::exists('gserver', ['nurl' => Strings::normaliseLink('http://' . $peer)])) {
+			if (DBA::exists('gserver', ['nurl' => 'http://' . $peer])) {
```
On normalizing we also remove trailing `/` and remove leading `www.`. This replacement does only half of the job.

Quix0r replied:
Please tell this @MrPetovan and not me. I never touched that code before, just added the `if()` blocks for checking if the URL is blocked. We should use the original URL for checking against block list. What if someone wants to block the `www.` part of an URL? Some have `www.friendica.tld` as their "homepage".

tek-aevl commented:
This is killing my power bill, I have 1000 of duplicate entry errors, because of this same issue.

MrPetovan commented:
Can you please share a sample of duplicate entry errors?

tek-aevl commented:
Ok, I forgot to respond to this sooner, 
<html>
<body>
<!--StartFragment--><html class=" js " style="box-sizing: border-box; font-family: sans-serif; text-size-adjust: 100%; font-size: 10px; -webkit-tap-highlight-color: rgba(0, 0, 0, 0);"><body id="top" class="mod-admin is-not-singleuser desktop-view modal-open" style="box-sizing: border-box; margin: 0px; font-family: &quot;Open Sans&quot;, sans-serif; font-size: 14px; line-height: 1.42857; color: rgb(204, 204, 204); background-color: rgb(0, 0, 0); padding-top: 110px; background-image: url(&quot;&quot;); background-size: cover; background-repeat: no-repeat; background-attachment: fixed; overflow: hidden; padding-right: 17px;"><main style="box-sizing: border-box; display: block;"><div class="container" style="box-sizing: border-box; padding-right: 15px; padding-left: 15px; margin-right: auto; margin-left: auto; width: 1170px;"><div class="row" style="box-sizing: border-box; margin-right: -15px; margin-left: -15px;"><div class="col-lg-7 col-md-7 col-sm-12 col-xs-12" id="content" style="box-sizing: border-box; position: relative; min-height: 1px; padding-right: 15px; padding-left: 15px; float: left; width: 682.5px;"><section class="sectiontop admin-content-wrapper" style="box-sizing: border-box; display: block;"><div id="logdetail" class="modal fade in" tabindex="-1" role="dialog" data-row-id="ev-118" style="box-sizing: border-box; opacity: 1; transition: opacity 0.15s linear 0s; position: fixed; inset: 0px; z-index: 1050; display: block; overflow: hidden auto; outline: 0px; padding-right: 17px;"><div class="modal-dialog modal-lg" role="document" style="box-sizing: border-box; position: relative; width: 1712.69px; margin: 30px auto; transform: translate(0px, 0px); transition: transform 0.3s ease-out 0s, -webkit-transform 0.3s ease-out 0s, -o-transform 0.3s ease-out 0s;"><div class="modal-content" style="box-sizing: border-box; position: relative; background-color: rgb(0, 0, 0); background-clip: padding-box; border: 1px solid rgba(0, 0, 0, 0.2); border-radius: 6px; box-shadow: rgba(0, 0, 0, 0.5) 0px 5px 15px; outline: 0px;"><div class="modal-header" style="box-sizing: border-box; padding: 15px; border-bottom: 1px solid rgb(229, 229, 229);"><h4 class="modal-title" style="box-sizing: border-box; margin: 0px; font-weight: 500; line-height: 1.42857; text-rendering: optimizelegibility; font-size: 18px; font-family: inherit; color: inherit;">Event details</h4></div><div class="modal-body" style="box-sizing: border-box; position: relative; padding: 15px;">

Date | Level | Context | Message
-- | -- | -- | --
2023-04-12T12:50:56Z | ERROR | worker | DB Error

<h3 class="event-data-header" style="box-sizing: border-box; margin: 20px 0px 10px; font-weight: 500; line-height: 1.1; text-rendering: optimizelegibility; font-size: 24px; font-family: inherit; color: inherit;">Data</h3><div class="event-source" style="box-sizing: border-box;"></div><div class="event-data" style="box-sizing: border-box;"><details open="" style="box-sizing: border-box; display: block; padding: 0.5em 0.5em 0px;"><summary style="box-sizing: border-box; display: list-item; cursor: pointer; font-weight: bold;">code</summary><pre style="box-sizing: border-box; overflow: auto; font-family: Menlo, Monaco, Consolas, &quot;Courier New&quot;, monospace; font-size: 13px; display: block; padding: 9.5px; margin: 0px 0px 10px; line-height: 1.42857; color: rgb(172, 172, 172); word-break: break-all; overflow-wrap: break-word; background-color: rgba(255, 255, 255, 0.05); border: 1px solid rgb(204, 204, 204); border-radius: 4px;">2006</pre></details><details open="" style="box-sizing: border-box; display: block; padding: 0.5em 0.5em 0px;"><summary style="box-sizing: border-box; display: list-item; cursor: pointer; font-weight: bold;">error</summary><pre style="box-sizing: border-box; overflow: auto; font-family: Menlo, Monaco, Consolas, &quot;Courier New&quot;, monospace; font-size: 13px; display: block; padding: 9.5px; margin: 0px 0px 10px; line-height: 1.42857; color: rgb(172, 172, 172); word-break: break-all; overflow-wrap: break-word; background-color: rgba(255, 255, 255, 0.05); border: 1px solid rgb(204, 204, 204); border-radius: 4px;">MySQL server has gone away</pre></details><details open="" style="box-sizing: border-box; display: block; padding: 0.5em 0.5em 0px;"><summary style="box-sizing: border-box; display: list-item; cursor: pointer; font-weight: bold;">callstack</summary><pre style="box-sizing: border-box; overflow: auto; font-family: Menlo, Monaco, Consolas, &quot;Courier New&quot;, monospace; font-size: 13px; display: block; padding: 9.5px; margin: 0px 0px 10px; line-height: 1.42857; color: rgb(172, 172, 172); word-break: break-all; overflow-wrap: break-word; background-color: rgba(255, 255, 255, 0.05); border: 1px solid rgb(204, 204, 204); border-radius: 4px;">GServer::insert, GServer::setFailure, GServer::detect, GServer::check, GServer::getID, GServer::add, UpdateServerPeers::execute, Worker::execFunction</pre></details><details open="" style="box-sizing: border-box; display: block; padding: 0.5em 0.5em 0px;"><summary style="box-sizing: border-box; display: list-item; cursor: pointer; font-weight: bold;">params</summary><pre style="box-sizing: border-box; overflow: auto; font-family: Menlo, Monaco, Consolas, &quot;Courier New&quot;, monospace; font-size: 13px; display: block; padding: 9.5px; margin: 0px 0px 10px; line-height: 1.42857; color: rgb(172, 172, 172); word-break: break-all; overflow-wrap: break-word; background-color: rgba(255, 255, 255, 0.05); border: 1px solid rgb(204, 204, 204); border-radius: 4px;">INSERT INTO `gserver` (`url`, `nurl`, `network`, `created`, `last_failure`, `failed`) VALUES ('https://32tbqwum0.activitypub-troll.cf', 'http://32tbqwum0.activitypub-troll.cf', 'unkn', '2023-04-12 12:50:56', '2023-04-12 12:50:56', 1)</pre></details></div></div></body></html><!--EndFragment-->
</body>
</html>






<html>
<body>
<!--StartFragment--><html class=" js " style="box-sizing: border-box; font-family: sans-serif; text-size-adjust: 100%; font-size: 10px; -webkit-tap-highlight-color: rgba(0, 0, 0, 0);"><body id="top" class="mod-admin is-not-singleuser desktop-view modal-open" style="box-sizing: border-box; margin: 0px; font-family: &quot;Open Sans&quot;, sans-serif; font-size: 14px; line-height: 1.42857; color: rgb(204, 204, 204); background-color: rgb(0, 0, 0); padding-top: 110px; background-image: url(&quot;&quot;); background-size: cover; background-repeat: no-repeat; background-attachment: fixed; overflow: hidden; padding-right: 17px;"><main style="box-sizing: border-box; display: block;"><div class="container" style="box-sizing: border-box; padding-right: 15px; padding-left: 15px; margin-right: auto; margin-left: auto; width: 1170px;"><div class="row" style="box-sizing: border-box; margin-right: -15px; margin-left: -15px;"><div class="col-lg-7 col-md-7 col-sm-12 col-xs-12" id="content" style="box-sizing: border-box; position: relative; min-height: 1px; padding-right: 15px; padding-left: 15px; float: left; width: 682.5px;"><section class="sectiontop admin-content-wrapper" style="box-sizing: border-box; display: block;"><div id="logdetail" class="modal fade in" tabindex="-1" role="dialog" data-row-id="ev-500" style="box-sizing: border-box; opacity: 1; transition: opacity 0.15s linear 0s; position: fixed; inset: 0px; z-index: 1050; display: block; overflow: hidden auto; outline: 0px; padding-right: 17px;"><div class="modal-dialog modal-lg" role="document" style="box-sizing: border-box; position: relative; width: 1712.69px; margin: 30px auto; transform: translate(0px, 0px); transition: transform 0.3s ease-out 0s, -webkit-transform 0.3s ease-out 0s, -o-transform 0.3s ease-out 0s;"><div class="modal-content" style="box-sizing: border-box; position: relative; background-color: rgb(0, 0, 0); background-clip: padding-box; border: 1px solid rgba(0, 0, 0, 0.2); border-radius: 6px; box-shadow: rgba(0, 0, 0, 0.5) 0px 5px 15px; outline: 0px;"><div class="modal-body" style="box-sizing: border-box; position: relative; padding: 15px;">

2023-04-09T15:03:00Z | ERROR | worker | DB Error
-- | -- | -- | --


<h3 class="event-data-header" style="box-sizing: border-box; margin: 20px 0px 10px; font-weight: 500; line-height: 1.1; text-rendering: optimizelegibility; font-size: 24px; font-family: inherit; color: inherit;">Data</h3><div class="event-source" style="box-sizing: border-box;"></div><div class="event-data" style="box-sizing: border-box;"><details open="" style="box-sizing: border-box; display: block; padding: 0.5em 0.5em 0px;"><summary style="box-sizing: border-box; display: list-item; cursor: pointer; font-weight: bold;">code</summary><pre style="box-sizing: border-box; overflow: auto; font-family: Menlo, Monaco, Consolas, &quot;Courier New&quot;, monospace; font-size: 13px; display: block; padding: 9.5px; margin: 0px 0px 10px; line-height: 1.42857; color: rgb(172, 172, 172); word-break: break-all; overflow-wrap: break-word; background-color: rgba(255, 255, 255, 0.05); border: 1px solid rgb(204, 204, 204); border-radius: 4px;">1062</pre></details><details open="" style="box-sizing: border-box; display: block; padding: 0.5em 0.5em 0px;"><summary style="box-sizing: border-box; display: list-item; cursor: pointer; font-weight: bold;">error</summary><pre style="box-sizing: border-box; overflow: auto; font-family: Menlo, Monaco, Consolas, &quot;Courier New&quot;, monospace; font-size: 13px; display: block; padding: 9.5px; margin: 0px 0px 10px; line-height: 1.42857; color: rgb(172, 172, 172); word-break: break-all; overflow-wrap: break-word; background-color: rgba(255, 255, 255, 0.05); border: 1px solid rgb(204, 204, 204); border-radius: 4px;">Duplicate entry 'http://39cqpqb9m.activitypub-troll.cf' for key 'nurl'</pre></details><details open="" style="box-sizing: border-box; display: block; padding: 0.5em 0.5em 0px;"><summary style="box-sizing: border-box; display: list-item; cursor: pointer; font-weight: bold;">callstack</summary><pre style="box-sizing: border-box; overflow: auto; font-family: Menlo, Monaco, Consolas, &quot;Courier New&quot;, monospace; font-size: 13px; display: block; padding: 9.5px; margin: 0px 0px 10px; line-height: 1.42857; color: rgb(172, 172, 172); word-break: break-all; overflow-wrap: break-word; background-color: rgba(255, 255, 255, 0.05); border: 1px solid rgb(204, 204, 204); border-radius: 4px;">GServer::insert, GServer::setFailure, GServer::detect, GServer::check, GServer::getID, GServer::add, UpdateServerPeers::execute, Worker::execFunction</pre></details><details open="" style="box-sizing: border-box; display: block; padding: 0.5em 0.5em 0px;"><summary style="box-sizing: border-box; display: list-item; cursor: pointer; font-weight: bold;">params</summary><pre style="box-sizing: border-box; overflow: auto; font-family: Menlo, Monaco, Consolas, &quot;Courier New&quot;, monospace; font-size: 13px; display: block; padding: 9.5px; margin: 0px 0px 10px; line-height: 1.42857; color: rgb(172, 172, 172); word-break: break-all; overflow-wrap: break-word; background-color: rgba(255, 255, 255, 0.05); border: 1px solid rgb(204, 204, 204); border-radius: 4px;">INSERT INTO `gserver` (`url`, `nurl`, `network`, `created`, `last_failure`, `failed`) VALUES ('https://39cqpqb9m.activitypub-troll.cf', 'http://39cqpqb9m.activitypub-troll.cf', 'unkn', '2023-04-09 15:03:00', '2023-04-09 15:03:00', 1)</pre></details></div></div></body></html><!--EndFragment-->
</body>
</html>

MrPetovan commented:
Can you please confirm what version of Friendica you are running and what pattern you set in your server block list related to this attack?

tek-aevl commented:
https://www.urbanmind.net/friendica 

Using PiHole blacklisting, with (\.|^)activitypub-troll\.cf$  and cloudflare dns firewall.

MrPetovan commented:
Ok, the `*.activitypub-troll.cf` pattern looks good and you seem to be running the 2023.01 version. I also see you have enabled some sort of CloudFlare DDoS protection which prevents JS and CSS files to load.

MrPetovan commented:
So, I have looked into the chain of calls you provided, and it should have been stopped in 4 different places given the error and your blocked server domain pattern:
- In `src/Worker/UpdateServerPeers.php:60` when we check for the existence of a GServer by normalized URL.
- In `src/Model/GServer.php:127` when we try to retrieve a GServer by normalized URL.
- In `src/Model/GServer.php:344` when we try to retrieve a GServer by normalized URL.
- In `src/Model/GServer.php:412` when we try to retrieve a GServer by normalized URL.

I don't have a good explanation for this.

I also note that we've added a number of checks against the server domain pattern block list in the upcoming version **2023.04** so updating to it should reduce your woes.

tek-aevl commented:
The CSS not loading, was not intended, but it helps.

I think I fixed the css not loading issue, it was hidden in my web shield settings.
